### ISSUE TITLE
Broken multi-column specified on a non root or body element

### ISSUE THREAD

**Issue by kwkbtr** [vivliostyle/vivliostyle.js#579]

**Comment by MurakamiShinyu** [comment]
> - **問題**：ルートあるいは body 以外の要素に指定した段組が正常に組版できない。
> - **メモ**：現在の段組の実装は、EPUB Adaptive Layout でのページ区画で指定する段組の実装がベースで、ルートまたは body 要素に指定された段組をページエリア全体の段組の扱いに変えてサポートしている。それ以外の要素での段組のプロパティはそのままブラウザに渡されるので、ページの中で部分的に使われたときに機能するが、複数ページに渡るときの分割処理などはできていない。

(from https://vivliostyle.github.io/vivliostyle_doc/ja/vivliostyle-user-group-vol2/shinyu/index.html)


- **Problem:** Columns specified for elements other than root or body cannot be formatted correctly.
- **Note:** The current implementation of columns is based on the implementation of columns specified in page partitions in EPUB Adaptive Layout, and supports columns specified in the root or body element by changing them to be treated as columns for the entire page area. Since the properties of the columns in other elements are passed to the browser as they are, they work when used in parts of the page, but they do not handle breaking up of the page when it spans multiple pages.

**Comment by danvim** [comment]
@MurakamiShinyu  Is this demonstrating the same issue? The problem here is a non-root multi-column container cannot flow into multi-column with `column-fill: auto`

https://vivliostyle.org/viewer/#src=data:,%3Cdiv%20class=%22col%22%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3Cp%3EParagraph%3C/p%3E%3C/div%3E&style=data:,/*%3Cviewer%3E*/%0A@page%20%7B%20size:%20a6%20landscape;%20%7D%0A/*%3C/viewer%3E*/%0A%0Ap%20%7B%0A%20%20margin:%200;%0A%7D%0A%0A.col%20%7B%0A%20%20columns:%202;%0A%20%20column-fill:%20auto;%0A%7D

**Comment by MurakamiShinyu** [comment]
@danvim Yes, same issue. The multi-column specified on `<div class="col">` does not work properly.

**Comment by danvim** [comment]
@MurakamiShinyu Thanks for confirming. Just a question, if I wish to have no multi-column main content and a section with multi-column, is the current way of achieving this to use EPUB Adaptive Layout rules?

**Comment by MurakamiShinyu** [comment]
@danvim In your case, the multi-column layout specified on the section will work if `column-fill: balance` is used. However, when the multi-column section is not fit in a page, auto page breaking may not work properly. We will fix this bug in the near future, but if you cant wait there will be some workarounds.

- use forced page break if auto page break may not work properly
- specify multi-column on the root element and use page floats (`float-reference: page; float: top`) for elements that need to be single-column.
- if your layout can be achieved with EPUB Adaptive Layout rules, that may be an option (note that the multi-column properties have to be specified on page partitions).

**Comment by MurakamiShinyu** [comment]
## Simple test samples

These simple tests show the problem that column/page breaking does not work properly and causes unexpected blank spaces or overflow when multi-column is applied to a div element.
 
- [cols2-test1.html](https://gist.github.com/MurakamiShinyu/6896d14a7b282c3af9421e66af004967#file-cols2-test1-html)
  - Test with Vivliostyle Viewer: https://vivliostyle.org/viewer/#src=https://gist.github.com/MurakamiShinyu/6896d14a7b282c3af9421e66af004967#file-cols2-test1-html
  - Result screenshot:  
    <img width="300" alt="Image" src="https://github.com/user-attachments/assets/428b683b-ce64-44ed-a648-ee0fb272e846" />
- [cols2-test2.html](https://gist.github.com/MurakamiShinyu/6896d14a7b282c3af9421e66af004967#file-cols2-test2-html)
  - Test with Vivliostyle Viewer: https://vivliostyle.org/viewer/#src=https://gist.github.com/MurakamiShinyu/6896d14a7b282c3af9421e66af004967#file-cols2-test2-html
  - Result screenshot:  
    <img width="300" alt="Image" src="https://github.com/user-attachments/assets/12620775-11c2-4a48-b472-9e731da13831" />

**Comment by MurakamiShinyu** [comment]
Fixed in PR #1571.

Test:
- https://vivliostyle-git-fix-issue579-vivliostyle.vercel.app/#src=https://gist.github.com/MurakamiShinyu/6896d14a7b282c3af9421e66af004967#file-cols2-test1-html
  
  <img width="325" height="659" alt="Image" src="https://github.com/user-attachments/assets/e96c62ea-2fea-4c5e-ae54-21077138ed28" />

- https://vivliostyle-git-fix-issue579-vivliostyle.vercel.app/#src=https://gist.github.com/MurakamiShinyu/6896d14a7b282c3af9421e66af004967#file-cols2-test2-html
  
  <img width="327" height="658" alt="Image" src="https://github.com/user-attachments/assets/552eaef9-b457-4a42-b39a-b41c642109d0" />

**Comment by MurakamiShinyu** [comment]
https://github.com/vivliostyle/vivliostyle.js/pull/1571#issuecomment-3460378490:

> ### Limitation in Safari (WebKit)
> > This commit enhances the page and column breaking functionality by utilizing the browser's built-in multi-column layout capabilities, that can handle nested multi-column layouts.
> 
> This works well in Chromium-based browsers and Firefox, however, unfortunately, Safari (WebKit) has problems with nested multi-column layouts that still prevent correct page/column breaking in multi-column elements inside the body element.
> 
> Related WebKit bugs:
> 
> * [[css-multicol-1] Nested multi-columns are laid out incorrectly](https://bugs.webkit.org/show_bug.cgi?id=269197)
> * [Nested multi-columns can cut off text](https://bugs.webkit.org/show_bug.cgi?id=269202)

**Comment by MurakamiShinyu** [comment]
## Mixed Multi-column layout Test

- [mixed-multicol-test.html](https://gist.github.com/MurakamiShinyu/6896d14a7b282c3af9421e66af004967#file-mixed-multicol-test-html)
  - Test with Vivliostyle Viewer:
  https://vivliostyle.vercel.app/#src=https://gist.github.com/MurakamiShinyu/6896d14a7b282c3af9421e66af004967#file-mixed-multicol-test-html
  - Result screenshot:
<img width="2176" height="696" alt="Image" src="https://github.com/user-attachments/assets/d20911a0-c9dd-4e76-96da-5290c9e7704d" />
